### PR TITLE
Stop ignoring errors when writing DataFrame to csv, parquet, json

### PR DESCRIPTION
**Pull request by andygrove**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/3800 and https://github.com/apache/arrow-datafusion/issues/3783

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/src/physical_plan/file_format/csv.rs [hunk 1]
```diff
@@ -252,7 +252,13 @@ pub async fn plan_to_csv(
...
                 tasks.push(handle);
             }
-            futures::future::join_all(tasks).await;
+            let results = futures::future::join_all(tasks).await;
+            for result in results {
+                let result =
+                    result.map_err(|e| DataFusionError::Execution(format!("{}", e)))?;
+                let _ =
+                    result.map_err(|e| DataFusionError::Execution(format!("{}", e)))?;
+            }
```
Comment (andygrove) [inline datafusion/core/src/physical_plan/file_format/csv.rs:261]: open to suggestions on a better implementation here

Comment (andygrove) [reply datafusion/core/src/physical_plan/file_format/csv.rs:261]: I suppose some tests would be good too ... I will add some to this PR later today

Comment (alamb) [reply datafusion/core/src/physical_plan/file_format/csv.rs:261]: I think we can use `try_for_each` from https://docs.rs/futures-util/0.3.24/futures_util/stream/trait.TryStreamExt.html#method.try_for_each

**Review by alamb — APPROVED** [review]
Seems like a good improvement to me -- I left a suggestion on how to  make this easier on the eyes but I ran out of time today to try it myself

File: datafusion/core/src/physical_plan/file_format/csv.rs (hunk 1 above)
Comment (andygrove) [reply datafusion/core/src/physical_plan/file_format/csv.rs]: That worked well. Thank you.

File: datafusion/core/src/physical_plan/file_format/csv.rs [hunk 2]
```diff
@@ -270,7 +270,12 @@ pub async fn plan_to_csv(
                 });
                 tasks.push(handle);
             }
-            futures::future::join_all(tasks).await;
+            futures::future::join_all(tasks)
+                .await
+                .into_iter()
+                .try_for_each(|result| {
+                    result.map_err(|e| DataFusionError::Execution(format!("{}", e)))?
```
Comment (alamb) [inline datafusion/core/src/physical_plan/file_format/csv.rs:277]: Nice 👍

**Review by alamb — COMMENTED** [review]
Thanks @andygrove

**Pull request merged**